File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/StrengthPower.ts
import { WrapDamage } from "../DataDefine/DamageInfo";
import { CalcDamageValueMessage } from "../Events/CalcDamageValueMessage";
import { OnDamageBefore } from "../Events/OnDamageBefore";
import FileHelper from "../FileHelper";
import { LocalizationProperty } from "../Gen/DB/Localization";
import { StringHelper } from "../StringHelper";
import { AbstractPower, PowerType } from "./AbstractPower";

export class StrengthPower extends AbstractPower{
    Name: string = LocalizationProperty.Read("能力-力量");
    Icon: string = FileHelper.FormatPath("Powers/strength.png");
    Id: string = "Strength";
    get Type(){return this.Amount >= 0 ? PowerType.Buff : PowerType.Debuff}
    GetDescription(...args: any[]): string {
        if(this.Amount >= 0){
            return StringHelper.FormatColorString(LocalizationProperty.Read("能力-力量(提升)")).format(this.Amount);
        }
        return StringHelper.FormatColorString(LocalizationProperty.Read("能力-力量(下降)")).format(this.Amount);
    }
    //TODO
    protected CalcDamageValue(msg: CalcDamageValueMessage) {
        super.CalcDamageValue(msg);
        msg.info.AddModify(new WrapDamage(this,this.Amount))
    }
    
    protected OnDamageBefore(msg: OnDamageBefore) {
        super.OnDamageBefore(msg);
        msg.info.AddModify(new WrapDamage(this,this.Amount))
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Power/Buff/StrengthPower.cs
using ModTheSpire.DataDefine;
using ModTheSpire.Event.Ids;
using TCG.Config;

namespace ModTheSpire.Power.Buff
{
    public class StrengthPower : AbstractPower
    {
        public override string Id { get; } = nameof(StrengthPower);
        public override string Name { get; } = LocalizationProperty.Read("能力-力量");
        public override string Icon { get; } = "Powers/strength.png";
        public override PowerType Type => Amount >= 0 ? PowerType.Buff : PowerType.Debuff;

        public override string GetDescription(object[] args)
        {
            return string.Format(LocalizationProperty.Read(this.Amount >= 0 ? "能力-力量(提升)" : "能力-力量(下降)"),this.Amount);
        }

        protected override void OnDamageBefore(OnDamageBeforeMessage msg)
        {
            base.OnDamageBefore(msg);
            msg.DamageInfo.AddModify(new WrapDamage(this.Owner, this.Amount));
        }

        protected override void CalcDamageValue(CalcDamageValueMessage msg)
        {
            base.CalcDamageValue(msg);
            msg.DamageInfo.AddModify(new WrapDamage(this.Owner, this.Amount));
        }
    }
}